### DISCUSSION TITLE
Repeater Component in set value on total input box , based on rate and quantity

### DISCUSSION THREAD

**Discussion by henilcodes** [filamentphp/filament#12625] [Help] [answered]
### Package

Form builder

### Package Version

v3

### How can we help you?

![Screenshot from 2024-05-04 07-23-40](https://github.com/filamentphp/filament/assets/99518388/c6027a0e-11a8-4cd6-a15e-c462105c4bdf)

I use Repeater in filament PHP.

When I change the quantity, I want the store total value in the Repeater in the total input field.

How To Set Total In Filament Repeater Field.

Example :

Service Details:

Rate: 6; Qty: 100; Total: 600

Rate: 10; Qty:100; Total: 1000

I want the service details in  Total in input Field to be auto auto-calculated based on Quantity and Rate.

Source Code BillResource  controller:

`<?php

class BillingResource extends Resource
{
    protected static ?string $model = Billing::class;

    protected static ?string $navigationIcon = 'heroicon-o-banknotes';

    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Forms\Components\Group::make()->schema([
                    Forms\Components\Section::make('Billing Details')->schema([
                        Forms\Components\TextInput::make('name')->label('Bill Number')->required()->default('B-' . random_int(100000, 999999))->disabled()->dehydrated()
                            ->maxLength(32)->unique(Billing::class, 'name', ignoreRecord: true),
                        Forms\Components\DatePicker::make('date')->label('Bill Date')->required(),

                        Forms\Components\Select::make('customer_id')->relationship('customer', 'name')->label('Customer')->searchable()->required()
                            ->afterStateUpdated(function ($state, Forms\Set $set) {
                                $customer = Customer::find($state)->toArray();
                                $set('mobile', $customer['mobile_1']);
                            })->reactive()->distinct(),
                        Forms\Components\TextInput::make('mobile')->required()->label('Mobile Number')->disabled()->dehydrated(),

                        Forms\Components\Select::make('vehicle_id')->relationship('vehicle', 'name')->label('Vehicle')->searchable()->required(),

                        Forms\Components\Select::make('gst_type')->label('GST Type')->options(['state_gst' => 'State GST', 'integrated_gst' => 'Integrated GST', 'none' => 'none'])->searchable()->required(),
                    ])->columns(2)->collapsible(),
                    Forms\Components\Section::make('Service Details')->schema([
                        static::getBillingServiceDetails(),
                    ])->columnSpanFull()->collapsible(),

                    Forms\Components\Section::make('Payment Details')->schema([

                        Forms\Components\Group::make()->schema([
                            Forms\Components\Select::make('payment_mode')->label('Payment Mode')->required()->searchable()->options([
                                'upi' => 'UPI',
                                'neft' => 'NEFT',
                                'cash' => 'Cash',
                            ]),

                        ]),
                        Forms\Components\Group::make()->schema([
                            Forms\Components\TextInput::make('net_total')->label('Net Total')->numeric()->disabled()->dehydrated()->suffix('₹')->minValue(0)->required()->default(0),
                            Forms\Components\TextInput::make('discount')->label('Discount')->numeric()->suffix('₹')->minValue(0)->required()->default(0)->afterStateUpdated(function (Get $get, Set $set) {
                                self::updateBillServiceQtyRateTotal($get, $set);
                            })->live(true),
                            Forms\Components\TextInput::make('grand_total')->label('Grand Total')->disabled()->dehydrated()->numeric()->suffix('₹')->minValue(0)->required()->default(0),
                        ]),

                        Forms\Components\Textarea::make('description')->label('Remark')->columnSpanFull(),

                    ])->columns(2)->collapsible(),

                ])->columnSpan(['lg' => fn (?Billing $record) => $record === null ? 4 : 4]),
            ])->columns(4);
    }

    public static function getBillingServiceDetails(): Repeater
    {
        return Repeater::make('billing_services')->relationship()->schema([

            Forms\Components\TextInput::make('name')->label('Description')->required()->columnSpan('2'),

            Forms\Components\Select::make('service_id')->relationship('service', 'name')->label('Service')->searchable()->required()->afterStateUpdated(function ($state, Forms\Set $set, Get $get) {
                if (isset($state)) {
                    $service = Service::findOrFail($state)->toArray();
                    $set('hsn_code', $service['hsn_code']);
                    $set('rate', $service['rate']);
                } else {
                    $set('hsn_code', '');
                    $set('rate', 0);
                }
            })->reactive()->distinct()->columnSpan(2),

            Forms\Components\TextInput::make('hsn_code')->label('HSN Code')->disabled()->dehydrated()->required(),

            Forms\Components\TextInput::make('rate')->numeric()->label('Rate')->disabled()->required()->dehydrated()->required(),

            Forms\Components\TextInput::make('qty')->numeric()->label('Quantity')->default(0)->required(),
            Forms\Components\TextInput::make('total')->numeric()->label('Total')->default(0)->required()->disabled()->dehydrated(),

        ])->deleteAction(
            fn (Action $action) => $action->requiresConfirmation(),
        )->defaultItems(1)->orderColumn('sr_no')->hiddenLabel()->columns('8')
            ->required()->live(true)->cloneable()->addActionLabel('add more')
            ->afterStateUpdated(function (Get $get, Set $set) {
                self::updateBillServiceQtyRateTotal($get, $set);
            });
    }

    public static function updateBillServiceQtyRateTotal(Get $get, Set $set)
    {

        $serviceDetails = collect($get('billing_services'));

        $subTotal = $serviceDetails->reduce(function ($subtotal, $serviceDetail) {

            if ($serviceDetail['rate'] >= 0 && $serviceDetail['qty'] >= 0) {

                $total = $serviceDetail['rate'] *  $serviceDetail['qty'];
                return $subtotal + $total;
            }
        }, 0);

        $grandTotal = $subTotal - $get('discount');

        $set('net_total', $subTotal);
        $set('grand_total', $grandTotal);
    }
}
`

**Comment by rmitesh** [comment, marked as answer]
Hi @HENILCODES , 

I have did same functionality in one of my project, I have updated your `getBillingServiceDetails()` function
hope that will helps you or let me know if you face any error.

```php
public static function getBillingServiceDetails(): Repeater
{
    return Repeater::make('billing_services')
        ->relationship()
        ->deleteAction(
            fn (Action $action) => $action->requiresConfirmation(),
        )
        ->defaultItems(1)
        ->orderColumn('sr_no')
        ->hiddenLabel()
        ->columns('8')
        ->required()
        ->live(true)
        ->cloneable()
        ->addActionLabel('add more')
        ->afterStateUpdated(function (Get $get, Set $set) {
            self::updateBillServiceQtyRateTotal($get, $set);
        })
        ->schema([
            Forms\Components\TextInput::make('name')
                ->label('Description')
                ->required()
                ->columnSpan('2'),

        Forms\Components\Select::make('service_id')
            ->relationship('service', 'name')
            ->label('Service')
            ->searchable()
            ->required()
            // changed this function
            ->afterStateUpdated(function ($state, Forms\Get $get, Forms\Set $set) {
                $service = Service::find($state);
                $set('hsn_code', $service?->hsn_code ?? '');
                $set('rate', $service?->rate ?? 0);

                $set('total', ($service?->rate * $get('qty') ?? 1));

            })
            ->reactive()
            ->distinct()
            ->columnSpan(2),

        Forms\Components\TextInput::make('hsn_code')
            ->label('HSN Code')
            ->disabled()
            ->dehydrated()
            ->required(),

        Forms\Components\TextInput::make('rate')
            ->numeric()
            ->label('Rate')
            ->disabled()
            ->required()
            ->dehydrated()
            ->required(),

        Forms\Components\TextInput::make('qty')
            ->required()
            ->numeric()
            ->label('Quantity')
            ->default(0)
            // added following methods
            ->dehydrated()
            ->live()
            ->afterStateUpdated(function ($state, Forms\Get $get, Forms\Set $set) {
                $set('total', ($get('rate') * $state ?? 0));
            })
            ->afterStateHydrated(function ($state, Forms\Get $get, Forms\Set $set) {
                $set('total', ($get('rate') * $state ?? 0));
            }),

        Forms\Components\TextInput::make('total')
            ->numeric()
            ->label('Total')
            ->default(0)
            ->required()
            ->disabled()
            ->dehydrated(),

    ]);
}
```

**Comment by henilcodes** [comment]
Thank you, Bro,  it is work 
My Problem is fixed,

**Comment by henilcodes** [comment]
Thank You @rmitesh
